src/canisters/initializer/src/lib.rs

```diff
@@ -46,59 +46,137 @@ fn registry() -> Principal {
 }
 
 #[derive(CandidType, serde::Deserialize, Clone, Copy)]
-struct InitialeDeposits {
-    vault: u128,
-    db: u128,
-    proxy: u128,
+struct CycleManagement {
+    initial_supply: u128,
+    refueling_amount: u128,
+    refueling_threshold: u128,
+}
+
+#[derive(CandidType, serde::Deserialize, Clone, Copy)]
+struct CycleManagements {
+    refueling_interval: u64,
+    vault_intial_supply: u128,
+    indexer: CycleManagement,
+    db: CycleManagement,
+    proxy: CycleManagement,
+}
+
+impl CycleManagements {
+    fn initial_supply(&self) -> u128 {
+        self.vault_intial_supply
+            + self.indexer.initial_supply
+            + self.db.initial_supply
+            + self.proxy.initial_supply
+    }
+}
+
+#[derive(CandidType, serde::Deserialize, Clone, Copy, PartialEq, Eq, Debug)]
+pub struct RefuelTarget {
+    pub id: Principal,
+    pub amount: u128,
+    pub threshold: u128,
 }
 
 #[update]
-async fn initialize(deposits: InitialeDeposits) -> InitializeOutput {
-    let deposits_total = deposits.vault + deposits.db + deposits.proxy;
+#[candid_method(update)]
+async fn initialize(deployer: Principal, cycles: CycleManagements) -> InitializeOutput {
+    let deposits_total = cycles.initial_supply();
     if deposits_total > ic_cdk::api::call::msg_cycles_accept128(deposits_total) {
-        panic!("Acceptable cycles are less than the specified.")
+        panic!("Acceptable cycles are less than the specified in parameters.")
     }
+
     let principal = ic_cdk::caller();
-    let vault = create_new_canister(deposits.vault).await.unwrap();
-    install_vault(&vault, &principal).await.unwrap();
-    after_install(&vault).await.unwrap();
-    register(principal, vault).await;
-    ic_cdk::println!(
-        "Vault of {:?} installed at {:?}",
-        principal.to_string(),
-        vault.to_string()
-    );
-    let db = create_new_canister(deposits.db).await.unwrap();
+
+    let vault = create_new_canister(cycles.vault_intial_supply)
+        .await
+        .unwrap();
+    let controllers = &vec![deployer, vault, ic_cdk::api::id()];
```

Comment:
Why must the vault be a controller of db and proxy?

Reply:
to get balance.
only a controller can call ic_cdk::api::management_canister::main::canister_status.

it can be replaced by adding a query function to get the balance for each canister.